File: Ruby/Deepspace/lib/Hangar.rb
#encoding:utf-8
require_relative 'HangarToUI'

module Deepspace

class Hangar

  attr_reader :maxElements, :weapons, :shieldBoosters

  def initialize(capacity)
    @maxElements = capacity
    @weapons = Array.new
    @shieldBoosters = Array.new
  end

  def self.newCopy(copy)
    Hangar.new(copy.maxElements)
  end

  def getUIversion
    HangarToUI.new(self)
  end

  def addWeapon(w)
    if spaceAvailable
      @weapons.push(w)
    end
    spaceAvailable
  end

  def addShieldBooster(s)
    if spaceAvailable
      @shieldBoosters.push(s)
    end
    spaceAvailable
  end

  def removeWeapon(w)
     @weapons.delete_at(w)
  end

   def removeShieldBooster(s)
    @shieldBoosters.delete_at(s)
  end

  def to_s
      return "MaxElements: #{@maxElements} Weapons: #{@weapons.join(", ")} ShieldBoosters: #{@shieldBoosters.join(", ")}"
  end

  private
  def spaceAvailable
    return @weapons.length + @shieldBoosters.length < @maxElements
  end


end

end

File: Deepspace_java/Deepspace/src/deepspace/Hangar.java

package deepspace;
import java.util.ArrayList;

/**
 *
 * @author victor
 */
public class Hangar {
    private int maxElements;
    private ArrayList<ShieldBooster> shieldBoosters;
    private ArrayList<Weapon> weapons;
    
    Hangar(int capacity){
        maxElements=capacity;
    }
    
    Hangar(Hangar copy){
        maxElements = copy.maxElements;
        shieldBoosters = copy.shieldBoosters;
        weapons = copy.weapons;
    }

    public int getMaxElements() {
        return maxElements;
    }

    public ArrayList<ShieldBooster> getShieldBoosters() {
        return shieldBoosters;
    }

    public ArrayList<Weapon> getWeapons() {
        return weapons;
    }
    
    HangarToUI getUIversion(){
        return new HangarToUI(this);
    }
    
    private boolean spaceAvailable(){
        return shieldBoosters.size() + weapons.size() < maxElements;
    }
    
    public boolean addWeapon(Weapon w){
        if( spaceAvailable() ){
            weapons.add(w);
            return true;
        } else
            return false;              
    }
    
    public boolean addShieldBooster(ShieldBooster w){
        if( spaceAvailable() ){
            shieldBoosters.add(w);
            return true;
        } else
            return false;              
    }
    
   public Weapon removeWeapon(int w){
       return weapons.remove(w);          
    }
   
    public ShieldBooster removeShieldBooster(int s){
       return shieldBoosters.remove(s);
    }
   
    
   
    
}
